### PR TITLE
unit tests for find by query params

### PR DESCRIPTION
**Pull request by tcrossrfid**
### What changes were proposed in this pull request?

Adds unit tests for find by query params - one test which looks at the string parameters
(nameLike, monikerLike, objectUrnLike, type) and another that looks at lastUpdated.
### How is this patch documented?

The code is self-explanatory.
### How was this patch tested?

The patch is itself a test.

### CODE REVIEW

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java [hunk 1]
```diff
@@ -108,8 +113,208 @@ public void getObjects() throws Exception {
...
+
+        queryParams.put(ObjectPersistenceService.MODIFIED_AFTER, fourthDate);
+        response = objectPersistenceService
+            .findByQueryParameters(accountUrn, queryParams);
+        assertTrue(response.size() == 0);
+
+    }
+
+    // used by findByQueryParametersStringParameters()
+    private void findByQueryParametersUtil() throws Exception
```
Comment (ghost) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java]: A better name is something like `populateQueryData`, something more specific to what is being done than the generic `Util`.

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java [hunk 2]
```diff
@@ -108,8 +113,208 @@ public void getObjects() throws Exception {
...
+    {
+
+        final UUID accountUuid = UuidUtil.getNewUuid();
+
+        ObjectEntity entityNameOneTypeOne = ObjectEntity.builder().accountId(accountUuid)
+            .objectUrn("objectUrnQueryParams01").name("name one").type("type one").build();
+
+        ObjectEntity entityNameTwoTypeOne = ObjectEntity.builder().accountId(accountUuid)
+            .objectUrn("objectUrnQueryParams02").name("name two").type("type one").build();
```
Comment (ghost) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java]: If you are using string like `objectUrnQueryParams01` in more than one location a `private static final String` style constant would be best.  Reusing a constant will reduce the possibilities for false success/failure.

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java [hunk 3]
```diff
@@ -108,8 +113,208 @@ public void getObjects() throws Exception {
...
+        response = objectPersistenceService
+            .findByQueryParameters(accountUrn, queryParams);
+        assertTrue(response.size() == 3);
+
+        queryParams.put(ObjectPersistenceService.MONIKER_LIKE, "moniker o");
+        response = objectPersistenceService
+            .findByQueryParameters(accountUrn, queryParams);
+        assertTrue(response.size() == 1);
+
+        queryParams.put(ObjectPersistenceService.MONIKER_LIKE, "moniker t");
```
Comment (ghost) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java:211]: Are these differences specific edge cases?  It seems like a duplication of effort to exercise the `MONIKER_LIKE` operator with different search strings that outwardly appear to accomplish the same result.  I would think tests for 
- all
- none
- specific subset would accomplish the same goal

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java [hunk 4]
```diff
@@ -108,8 +113,208 @@ public void getObjects() throws Exception {
...
     @Test
-    public void findByQueryParameters() throws Exception {
-        // TODO to.. do...
+    public void findByQueryParametersStringParameters() throws Exception
+    {
+        findByQueryParametersUtil();
+
+        Map<String, Object> queryParams = new HashMap<>();
+
+        queryParams.put(ObjectPersistenceService.OBJECT_URN_LIKE, "objectUrnQueryParams");
```
Comment (ghost) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java:143]: It seems this can be accomplished with less work and cover the boundary/edge conditions and the common results, see comment below.

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java [hunk 5]
```diff
@@ -108,8 +113,208 @@ public void getObjects() throws Exception {
...
+
+        Long secondDate = new Date().getTime();
+        Thread.sleep(DELAY_BETWEEN_LAST_MODIFIED_DATES);
+
+        ObjectEntity secondObject = ObjectEntity.builder().accountId(accountUuid)
+            .objectUrn("objectUrnLastModTest2").name("last mod test 2").type("anythingIsFine").build();
+        this.objectRepository.save(secondObject);
+
+        Long thirdDate = new Date().getTime();
+        Thread.sleep(DELAY_BETWEEN_LAST_MODIFIED_DATES);
```
Comment (ghost) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java:255]: There has to be a better way.  Introducing `Thread.sleep()` creates unwanted delays in UT completion.  We should be able to populate the integration test database with canned data and query it.

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java (hunk 1 above)
Comment (tcrossrfid) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java]: Done.

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java (hunk 2 above)
Comment (tcrossrfid) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java]: Done. Most of those strings are only used once, but now they're all string constants at the top of the file (the naming strings, not the querying strings), so more readily hackable.

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java (hunk 3 above)
Comment (tcrossrfid) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java:211]: That's all I'm testing per string, plus verifying that it's doing an actual startsWith, thus "oniker". OK, there are a couple of places where I'm testing two different subsets. I can remove those if you want.

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java (hunk 5 above)
Comment (tcrossrfid) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java:255]: I'm not thinking of it - anything else would involve doing something evil to directly hack lastModifiedTimestamp. On the bright side, I've cut the delays down to 10 ms, so the additional cost is now 30 ms. Still working fine. Canned data seems to me a) more complicated and b) in violation of the unit test's independence as a piece of code.

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java (hunk 3 above)
Comment (ghost) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java:211]: I try to live by "As much as is necessary and no more", with apologies to Einstein.  What we're after is completeness and coverage.  Often you can achieve 100% line and/or branch coverage and not hit all the edge cases.  If there is a duplication then leave th eduplicate off.  Less code is less that needs to be "cared for and fed".

File: src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java (hunk 5 above)
Comment (ghost) [inline src/test/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceServiceTest.java:255]: There is a way to populate the database from a canned file. When that gets developed by someone we should revisit this.
